src/main/java/at/ac/tuwien/damap/rest/dmp/mapper/DmpDOMapper.java

```diff
@@ -288,46 +298,39 @@ public Dmp mapDOtoEntity(DmpDO dmpDO, Dmp dmp, MapperService mapperService) {
             });
         }
 
-        //remove all existing Host objects, that are not included in the DO anymore
+        // remove all existing Host objects, that are not included in the DO anymore
         List<Host> hostList = dmp.getHostList();
         List<Host> hostListToRemove = new ArrayList<>();
         hostList.forEach(host -> {
-            Optional<RepositoryDO> hostDOOptional = dmpDO.getRepositories().stream().filter(hostDO ->
-                    hostDO.getId() != null &&
-                            hostDO.getId().equals(host.id)).findFirst();
+            boolean keep = false;
+
             if (Repository.class.isAssignableFrom(host.getClass())) {
-                if (hostDOOptional.isEmpty()) {
-                    hostListToRemove.add(host);
-                }
+                keep &= dmpDO.getRepositories().stream().anyMatch(
+                        hostDO -> hostDO.getId() != null &&
+                                hostDO.getId().equals(host.id));
             }
-        });
-        hostList.forEach(host -> {
-            Optional<StorageDO> hostDOOptional = dmpDO.getStorage().stream().filter(hostDO ->
-                    hostDO.getId() != null &&
-                            hostDO.getId().equals(host.id)).findFirst();
             if (Storage.class.isAssignableFrom(host.getClass())) {
-                if (hostDOOptional.isEmpty()) {
-                    hostListToRemove.add(host);
-                }
+                keep &= dmpDO.getStorage().stream()
+                        .anyMatch(hostDO -> hostDO.getId() != null &&
+                                hostDO.getId().equals(host.id));
             }
-        });
-        hostList.forEach(host -> {
-            Optional<ExternalStorageDO> hostDOOptional = dmpDO.getExternalStorage().stream().filter(hostDO ->
-                    hostDO.getId() != null &&
-                            hostDO.getId().equals(host.id)).findFirst();
+
             if (ExternalStorage.class.isAssignableFrom(host.getClass())) {
-                if (hostDOOptional.isEmpty()) {
-                    hostListToRemove.add(host);
-                }
+                keep &= dmpDO.getExternalStorage().stream()
+                        .anyMatch(hostDO -> hostDO.getId() != null &&
+                                hostDO.getId().equals(host.id));
+            }
+
+            if (!keep) {
+                hostListToRemove.add(host);
             }
```

Comment:
Always true, since &= for booleans is the same as &&. keep is initialized as false, therefore keep stays false.

Reply:
good catch, thank you! it should be `|=` such that if it is found once, we keep it.